### DISCUSSION TITLE
Feature Request: Add support for DuckDB

### DISCUSSION THREAD

**Discussion by DeflateAwning** [diesel-rs/diesel#3983] [Ideas]
I see that you have support for SQLite, but I think it would be awesome if you added support for DuckDB, an alternative OLAP database that looks and feels very similar to SQLite.

It already has a Rust API, and has [`r2d2` support](https://github.com/sfackler/r2d2) (not sure if that's really helpful, but figured it was worth mentioning).

**Comment by weiznich** [comment]
We do not plan to add any more backends/database connection implementations to diesel itself anymore. We are already at the limit on what we can maintain as project therefore we just don't have the capacity for more connection implementations.

Fortunately it's possible to just implement such a custom backend/connection outside of diesel, with essentially the same level of functionality as it would have as if it were implemented in diesel itself. You might want to consult the [documentation](https://docs.diesel.rs/2.1.x/diesel/connection/trait.Connection.html#implement-support-for-an-unsupported-database-system) on this topic or checkout existing third party implementations like [diesel-oci (oracle database)](https://github.com/GiGainfosystems/diesel-oci) or [rsfbclient-diesel (firebird database)](https://github.com/fernandobatels/rsfbclient/tree/master/rsfbclient-diesel). If you choose to go down that route we are happy to provide some more detailed pointers about how to approach such an implementation.

**Reply by liubao68** [reply]
Does Diesel consider list non-official supported implementations in README so that users can find them easier? And provide some test infrastructures to help non-official implementations to pass all Diesel test cases?

**Reply by weiznich** [reply]
@liubao68 Such third party extensions are listed at the [diesel home page](https://diesel.rs) at the bottom of the page. We are happy to list any diesel related extensions there. 

As for the test suite: Diesel has tests in several parts of the code base: Once as unit tests in the main crate, once as doc tests in the main crate and once as interaction tests in a separate crate. For the first two cases it's obviously hard to use them outside diesel, but at least the unit tests are testing mostly connection specific internals which are not useful for third party backends. 
The integration tests are in a separate not published crate in the workspace. They are kind of well separated so I suppose it shouldn't be that hard to either copy them over or include them as git submodule in a third party backend. You would still need to add the necessary setup for that as patch on top, but that shouldn't be too hard. 


For the out of tree extensions I wrote, I didn't bother with that but instead wrote a smaller test suite there to only test relevant stuff there, as the integration tests also test a lot of general stuff that's not relevant for third party backends.

**Comment by nimdeveloper** [comment]
I just started developing a crate for that, it's still in development but you check it here, and contribute if needed.
https://github.com/nimdeveloper/better-duck
